cookbook/basic_rag_example/pinecone_embeddings.py

```diff
@@ -0,0 +1,151 @@
+        num_splits = len(tokens) // max_tokens + 1
+        words_per_split = len(words) // num_splits
+        split_texts = [
+            " ".join(words[i : i + words_per_split])
+            for i in range(0, len(words), words_per_split + 1)
+        ]
+        return split_texts
+
+    for i, row in df.iterrows():
+        text = row["Text"]
+        tokens = encoder.encode(text)
+        if len(tokens) > 1000:
+            split_articles += split_text(text, tokens, 1000)
+            df.drop(i, inplace=True)
+
+    # Long texts which were dropped from the dataframe are now readded.
+    df = pd.concat(
+        [df, pd.DataFrame(split_articles, columns=["Text"])], ignore_index=True
+    )
+
+    return df
+
+
+def embed_data(df, batch_size):
+    """Embeds a dataframe of texts and stores it in a new column."""
+    embeddings = []
+    # ada 02 supports ~8000 tokens and 8 articles fit in the max batch size
+    for start in range(0, len(df), batch_size):
+        end = start + batch_size
+        batch = df["Text"][start:end].tolist()
+        embeddings_response = chat.client.embeddings.create(
+            model=embeddings_model, input=batch
+        )
+        embeddings += [datum.embedding for datum in embeddings_response.data]
+    df["embeddings"] = embeddings
+
+    return df
+
+
+filename = "news_article_dataset.csv"
+# Make reruns efficient by preprocessing data before saving to csv
+if not os.path.exists("cookbook/basic_rag_example/news_article_dataset.csv"):
+    url = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
+    df = load_data(url)
+    df = embed_data(df, 8)  # ada 02 supports ~8000 tokens, our max_tokens is 1000
+    df.to_csv(f"{os.getcwd()}/cookbook/basic_rag_example/{filename}")
+else:
+    df = pd.read_csv(f"{os.getcwd()}/cookbook/basic_rag_example/{filename}")
+
+
+pc = Pinecone(api_key="YOUR_API_KEY")
+existing_indexes = pc.list_indexes().names()
+index_name = "news-articles"
+if index_name not in existing_indexes:
+    pc.create_index(
+        name=index_name,
+        dimension=1536,
+        metric="cosine",
+        spec=ServerlessSpec(
+            cloud="aws",
+            region="us-west-2",
+        ),
+    )
+index = pc.Index(index_name)
+
+
+batch_size = 300  # Pinecone can't handle too large of an upsert request
+for start in range(0, len(df), batch_size):
+    end = min(start + batch_size, len(df))
+    batch = df[start:end]
+    vectors = []
+    for i, row in batch.iterrows():
+        embeddings = row["embeddings"]
+        # Embeddings need to be parsed if they've been read from a csv
+        if isinstance(embeddings[0], str):
+            embeddings = ast.literal_eval(embeddings)
+        vectors.append({"id": str(i), "values": embeddings})
+    index.upsert(vectors=vectors, namespace="articles")
+
+
+class NewsPrompt(Prompt):
+    """
+    Here are 4 relevant sections of news articles relevant to this topic: {topic}.
```

Comment:
update 4 here to be `{num_articles}` and then add a property that is the length of the articles list